mongo/integration/client_test.go

```diff
@@ -603,6 +600,99 @@ func TestClient(t *testing.T) {
 		assert.Equal(t, 0, closed, "expected no connections to be closed")
 	})
 
+	mt.Run("RTT90 is monitored", func(mt *mtest.T) {
+		if testing.Short() {
+			t.Skip("skipping integration test in short mode")
+		}
+
+		// Reset the client with a dialer that delays all network round trips by 300ms and set the
+		// heartbeat interval to 100ms to reduce the time it takes to collect RTT samples.
+		mt.ResetClient(options.Client().
+			SetDialer(newSlowConnDialer(300 * time.Millisecond)).
+			SetHeartbeatInterval(100 * time.Millisecond))
+
+		// Assert that RTT90s are eventually >300ms.
+		topo := getTopologyFromClient(mt.Client)
+		assert.Soon(mt, func() {
+			for {
+				time.Sleep(100 * time.Millisecond)
+
+				// Wait for all of the server's RTT90s to be >300ms.
+				done := true
+				for _, desc := range topo.Description().Servers {
+					server, err := topo.FindServer(desc)
+					assert.Nil(mt, err, "FindServer error: %v", err)
+					if server.RTT90() <= 300*time.Millisecond {
+						done = false
+					}
+				}
+				if done {
+					return
+				}
+			}
+		}, 10*time.Second)
+	})
+
+	// Test that if Timeout is set and the RTT90 is greater than the remaining timeout for an operation, the
+	// operation is not sent to the server, fails with a timeout error, and no connections are closed.
+	mt.Run("RTT90 used to prevent sending requests", func(mt *mtest.T) {
+		if testing.Short() {
+			t.Skip("skipping integration test in short mode")
+		}
+
+		// Assert that we can call Ping with a 250ms timeout.
+		ctx, cancel := context.WithTimeout(context.Background(), 250*time.Millisecond)
+		defer cancel()
+		err := mt.Client.Ping(ctx, nil)
+		assert.Nil(mt, err, "Ping error: %v", err)
+
+		// Reset the client with a dialer that delays all network round trips by 300ms, set the
+		// heartbeat interval to 100ms to reduce the time it takes to collect RTT samples, and
+		// set a Timeout of 0 (infinite) on the Client to ensure that RTT90 is used as a sending
+		// threshold.
+		tpm := monitor.NewTestPoolMonitor()
+		mt.ResetClient(options.Client().
+			SetPoolMonitor(tpm.PoolMonitor).
+			SetDialer(newSlowConnDialer(300 * time.Millisecond)).
+			SetHeartbeatInterval(100 * time.Millisecond).
+			SetTimeout(0))
+
+		// Assert that RTT90s are eventually >300ms.
+		topo := getTopologyFromClient(mt.Client)
+		assert.Soon(mt, func() {
+			for {
+				time.Sleep(100 * time.Millisecond)
+
+				// Wait for all of the server's RTT90s to be >300ms.
+				done := true
+				for _, desc := range topo.Description().Servers {
+					server, err := topo.FindServer(desc)
+					assert.Nil(mt, err, "FindServer error: %v", err)
+					if server.MinRTT() <= 300*time.Millisecond {
+						done = false
+					}
```

Comment:
We should break the outer loop when we go through each server in `topo.Description().Servers` and find that _each_ server has a RTT90() greater than 300ms. Returning here would mean that we would break once we found a single server with an RTT90 greater than 300ms, which seems incorrect to me.

I _was_ using `MinRTT` here when I should've been using `RTT90`, though. Updated.

Reply:
I mean to write `break` not `return`, mb!  I did this on my phone :( 

Once you find just one false value in the loop you can break and let the outer loop resume.  You don't have to look at all the servers.